File: Problems/0097. Interleaving String/solution2.py
# DP Top-Down
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        if len(s1) + len(s2) != len(s3):
            return False

        dp = {}
        
        def dfs(i, j):
            if (i, j) in dp:
                return dp[(i, j)]
            if i + j == len(s3):
                return True

            if len(s1) > i and len(s2) > j and s1[i] == s3[i + j] and s2[j] == s3[i + j]:
                dp[(i, j)] = dfs(i + 1, j) or dfs(i, j + 1)
            elif len(s1) > i and s1[i] == s3[i + j]:
                dp[(i, j)] = dfs(i + 1, j)
            elif len(s2) > j and s2[j] == s3[i + j]:
                dp[(i, j)] = dfs(i, j + 1)
            else:
                dp[(i, j)] = False

            return dp[(i, j)]

        return dfs(0, 0)

File: Problems/0105. Construct Binary Tree From Preorder And Inorder Traversal/solution2.kt
# Same as solution 1 but a bit broken down
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        inorderMap = {val: i for i, val in enumerate(inorder)}

        def recurse(preL, preR, inL, inR):
            if preL > preR or inL > inR:
                return None

            rootVal = preorder[preL]
            root = TreeNode(rootVal)
            inorderIndex = inorderMap[rootVal]

            LeftHalf = inorderIndex - inL

            root.left = recurse(preL + 1, preL + LeftHalf, inL, inorderIndex - 1)
            root.right = recurse(preL + 1 + LeftHalf, preR, inorderIndex + 1, inR)

            return root


        return recurse(0, len(preorder) - 1, 0, len(inorder) - 1)
